proto_analyzer/Hypothesis.fs

```diff
@@ -0,0 +1,45 @@
+﻿module Hypothesis
+
+// TODO pattern note pitch values are standard midi notes
+// TODO pattern dump offset 0x216: normal pattern len
+// TODO pattern dump offset 0x20f: normal pattern spd, 0x30 is 1/8, 0x3C is 1x
+// TODO tracks are store sequentially in a pattern - Track1-4, then FX, then CV;
+// TODO every track can have its own pattern scale settings.
+// TODO for both trigger and note pitch data, payloads are grouped in 7-byte blocks, separated by flipping 2a 55 2a 55s (acting as sync bytes?)
+//      00 in place of 2a/55 means no data in the block
+//      so, the raw bytes:  01 40 09 55 41 01 44 09 45 01 48 2a 09 49 01 4c 09 4d 01 55 50 09 51 01 54 09 55 2a
+//      are really:         01 40 09    41 01 44 09 45 01 48    09 49 01 4c 09 4d 01    50 09 51 01 54 09 55
```

Comment:
This sounds like elektron's 7bit encoding method. To pack an array of bytes in to an array of 7 bit values, they take 7 bytes at a time. Each byte has the MSB removed, this MSB is stored in an extra byte for each of the 7 values.

So your 7bit data stream looks like:
[ MSBs ]  [ 7bitValue ]  [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]   [ 7bitValue ]  (repeat)

"How the 7 bit encoding is generated
As a SysEx message only allows 7 bit data, 8 bit data blocks are converted as follows:
The first data (byte) contains the MSB of the following, up to, 7 data, data[0..6]. Bit 6 is the MSB bit of data[0] and bit 0 is the MSB of data[6]. 7 bytes encodes to 8 data. 8 bytes encodes to 10 data."

Reply:
I should add. Please refer to the mnm pdf i uploaded explaining packing and encoding.

Reply:
thanks, that's very interesting. I'll read up the mnm manual to follow up.
the MSBs probably aren't used here though, as the truth table already covers all the information for triggers.
also, please note:
```
0x55 = 0101 0101
0x2a = 0010 1010
```
So they add up to 0x7F (and do not chance however I arm the triggers) -- more like a synchronization pattern -- in case some bytes are dropped, it still recovers "on the beat".
My reasoning goes the same with the 01-09 pattern.
Another explaination would be, that all the effective bytes (0x40, 0x41 etc) have the MSB toggled, and the 01-09 patterns do not. If so, because each block contains 7 bytes, the pattern naturally shuffles like `1010101   0101010  1010101  0101010 ... `.

it surprised me that the 7bit and 8bit COM protocols are compatible, so the monitor program works without even touching the width and stop bit settings. :p